### ISSUE TITLE
Use bitmasks for timed traffic light states?

### ISSUE THREAD

**Issue by aubergine10** [krzychu124/Cities-Skylines-Traffic-Manager-President-Edition#470]
[`TLM/TLM/TrafficLight/Impl/CustomSegmentLights.cs`](https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/blob/master/TLM/TLM/TrafficLight/Impl/CustomSegmentLights.cs)

There's a bunch of similar functions such as: `IsAnyGreen()`, `IsAnyInTransition()`, etc...

Each one has the pattern:

```csharp
public bool Is%Whatever%() {
    foreach (KeyValuePair<ExtVehicleType, ICustomSegmentLight> e in CustomLights) {
        if (e.Value.Is%Whatever%()) {
            return %bool%;
        }
    }

    return %!bool%;
}
```

I don't know how often those methods are called, but it seems a crazy amount of work to be doing just to check if the traffic lights are in given state for given direction.

Surely we could use some bit masks here?

First bit mask would define how many lights there are for a segment node, and (potentially) in which lanes they appear. For example, this street:

![image](https://user-images.githubusercontent.com/1386719/62019903-cb096480-b1b8-11e9-9448-16bc6902788a.png)

Would be represented as `0 00 0011 10 0`. Let's call that `LightBits`. (I'm just making up names here for easier reference in this issue ticket, I have no idea what proper naming convention would be).

So just by the bit mask we now know where all the traffic lights are for this segment node.

Next, a bit mask is for red/green state, it would basically be same as above but the bits where lights are (`1`'s above) would be either `1` for green or `0` for red. Let's call this one `ColorBits`.

Then we need a bit mask to denote lights in transition between state, again, same sort of thing. `1` = transitioning, `0` = not transitioning. We'll call this one `TransitionBits`. We also gain an advantage here that we know what we are transitioning from, because we can look at the `LightBits` (eg. if light bit is `1` and transition bit is `1` we know that we're transitioning from green to red).

And we can multiplex all these bit masks in to a single word, if we set a limit on the number of lanes we can handle. Let's say the limit is 20 lanes on a segment. I don't think I've ever seen a segment bigger than that, and if someone has one I think it's fair to say "for performance reasons, we don't support anything bigger than 20 lanes". Anyway, we then fit our three bitmasks in to a single 64bit word, with 4 bits left over for... whatever.

But we also need directions - left, forward, right. That's just another 3 bitmasks multiplexed in to a 64 bit word (4 bits spare).

Each of the mentioned methods then become much leaner (from CPU - and memalloc/gc? - perspective) as we only need to do some bitwise math using two 64bit words (or just one if ignoring direction, eg. `IsAnyInTransition` doesn't care about direction).

It also means we can ditch a load of methods on the lower level class (`ICustomSegmentLight`).

**Comment by kvakvs** [comment]
I don't think we need to bit-pack memory saving for thing such as traffic lights. Even if it was a traffic light on 50% nodes, that's 32K lights max, and even if each light had 1kb state that'd be 32MB state (much exaggerated). A performant and easy to understand structure would be more valuable. That is having all related stuff per traffic light grouped in a machine word (8 bytes) or a few words, and avoid extra allocations as much as we can by using arrays or something.

Though we can bit-pack when saving if we come up with an error-resistant encoding which won't possibly break in say next 5 years.

**Comment by kvakvs** [comment]
For the situations when traffic light can be simple (all directions) or one per direction, or combined, the state class should have that all logic hidden in properties with debug time checks. Or maybe use a type-safe generic like `Either<Left, Right>` which can only be `Left` type or `Right` and not both at once.
